downstairs/src/extent_inner_sqlite.rs

```diff
@@ -0,0 +1,1741 @@
+        });
+
+        let mut hashes_to_write =
+            Vec::with_capacity(write.block_contexts.len());
+        for (i, ctx) in write.block_contexts.iter().enumerate() {
+            if writes_to_skip.contains(&i) {
+                hashes_to_write.push(None);
+                continue;
+            }
+
+            // TODO it would be nice if we could profile what % of time we're
+            // spending on hashes locally vs sqlite
+            let on_disk_hash =
+                integrity_hash(&[&write.data[i * block_size as usize..]
+                    [..block_size as usize]]);
+            let offset = i as u64 + write.offset.value;
+
+            self.set_block_context(&DownstairsBlockContext {
+                block_context: *ctx,
+                block: offset,
+                on_disk_hash,
+            })?;
+
+            // Worth some thought: this could happen inside
+            // tx_set_block_context, if we passed a reference to dirty_blocks
+            // into that function too. This might be nice, since then a block
+            // context could never be set without marking the block as dirty.
+            // On the other paw, our test suite very much likes the fact that
+            // tx_set_block_context doesn't mark blocks as dirty, since it
+            // lets us easily test specific edge-cases of the database state.
+            // Could make another function that wraps tx_set_block_context
+            // and handles this as well.
+            self.dirty_blocks.insert(offset as usize, None);
+            hashes_to_write.push(Some(on_disk_hash));
+        }
+        tx.commit()?;
+
+        cdt::extent__write__sqlite__insert__done!(|| {
+            (job_id.0, self.extent_number.0, num_blocks)
+        });
+
+        // PERFORMANCE TODO:
+        //
+        // Something worth considering for small writes is that, based on
+        // my memory of conversations we had with propolis folks about what
+        // OSes expect out of an NVMe driver, I believe our contract with the
+        // upstairs doesn't require us to have the writes inside the file
+        // until after a flush() returns. If that is indeed true, we could
+        // buffer a certain amount of writes, only actually writing that
+        // buffer when either a flush is issued or the buffer exceeds some
+        // set size (based on our memory constraints). This would have
+        // benefits on any workload that frequently writes to the same block
+        // between flushes, would have benefits for small contiguous writes
+        // issued over multiple write commands by letting us batch them into
+        // a larger write, and (speculation) may benefit non-contiguous writes
+        // by cutting down the number of sqlite transactions. But, it
+        // introduces complexity. The time spent implementing that would
+        // probably better be spent switching to aio or something like that.
```

Comment:
hah, i see my past self was on the same train of thought as me now. though i think i'm more in favor of doing that kind of thing upstairs closer to the guest these days